data/build_probe_4_data.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2020 Google LLC
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     https://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Scripts For Building Probe 4 (Movie/Tag-to-Movie Recommendations)."""
+
+import json
+import random
+
+from absl import app
+from absl import flags
+from absl import logging
+import numpy as np
+import tensorflow.compat.v1 as tf
+from tqdm import tqdm
+from trainer import constants
+
+FLAGS = flags.FLAGS
+flags.DEFINE_integer("random_seed", 1, "seed for random movie selection. Choose"
+                     + "-1 for a randomly picked seed")
+flags.DEFINE_integer("probe_min_pop", 30, "minimum poularity to be in probe")
+flags.DEFINE_integer("popular_min_pop", 138, "minimum popularity to be"
+                     + " considered a popular movie")
+
+
+def main(_):
+  """generate probe 4 data from movielens tags."""
+  logging.info("generating probe_4.tsv")
+  # set random seed for picking random movies
+  if FLAGS.random_seed != -1:
+    random.seed(FLAGS.random_seed)
+
+  with tf.io.gfile.GFile(constants.MATRIX_PATHS["pmi_matrix"], "rb") as f:
+    pmi_matrix = np.load(f)
+
+  with tf.io.gfile.GFile(constants.MATRIX_PATHS["movie_ids"], "r") as f:
+    movie_ids = json.load(f)
+
+  # define "popular" set as movie which appear in over FLAGS.popular_min_pop
+  # user sequences
+
+  popular_movies = [x.lower() for x in movie_ids["all_movies"]
+                    if movie_ids["popularity"][x] >= FLAGS.popular_min_pop]
+
+  logging.info("popular movies: filtered %d movies where popularity > %d",
+               len(popular_movies), FLAGS.popular_min_pop)
+
+  # define "filtered" set as movie which appear in over FLAGS.probe_min_pop
+  # user sequences
+
+  filtered_movies = [x.lower() for x in movie_ids["all_movies"]
+                     if movie_ids["popularity"][x] >= FLAGS.probe_min_pop]
+
+  logging.info("filtered movies: filtered %d movies where popularity > %d",
+               len(filtered_movies), FLAGS.probe_min_pop)
+
+  # lowercase movie titles to match ml_tags data
+  movie_ids["all_movies"] = [x.lower() for x in movie_ids["all_movies"]]
+  ids = list(range(len(movie_ids["all_movies"])))
+  movie_ids["movie_to_id"] = dict(zip(movie_ids["all_movies"], ids))
+  movie_ids["id_to_movie"] = dict(zip(ids, movie_ids["all_movies"]))
+
+  def intersection(list1, list2):
+    return [x for x in list1 if x in list2]
+
+  filtered_set = set(filtered_movies)
+  def get_related_movies(movie, k=5):
```

Comment:
Is there an opportunity to share get_related_movies function between probe 1 and probe 4 code?

Reply:
Yeah I changed it. I'm just importing it from build_probe_1_data now.